WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/Catalog Visibility/ProductCatalogVisibilityViewController.swift

```diff
@@ -0,0 +1,234 @@
+        tableView.register(SwitchTableViewCell.loadNib(), forCellReuseIdentifier: SwitchTableViewCell.reuseIdentifier)
+        tableView.register(BasicTableViewCell.loadNib(), forCellReuseIdentifier: BasicTableViewCell.reuseIdentifier)
+        tableView.register(TwoColumnSectionHeaderView.loadNib(), forHeaderFooterViewReuseIdentifier: TwoColumnSectionHeaderView.reuseIdentifier)
+
+        tableView.dataSource = self
+        tableView.delegate = self
+
+        tableView.backgroundColor = .listBackground
+        tableView.removeLastCellSeparator()
+    }
+
+    func configureTableViewFooter() {
+        /// `tableView.tableFooterView` can't handle a footerView that uses autolayout only.
+        /// Hence the container view with a defined frame.
+        let footerContainer = UIView(frame: CGRect(x: 0, y: 0, width: Int(tableView.frame.width), height: 1))
+        let footerView = FootnoteTableFooterView.instantiateFromNib() as FootnoteTableFooterView
+        footerView.footnoteLabel.text = NSLocalizedString("This setting determines which shop pages products will be listed on.",
+                                                          comment: "Footer text in Product Catalog Visibility")
+        tableView.tableFooterView = footerContainer
+        footerContainer.addSubview(footerView)
+    }
+}
+
+// MARK: - UITableViewDataSource Conformance
+//
+extension ProductCatalogVisibilityViewController: UITableViewDataSource {
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return sections.count
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return sections[section].rows.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let row = sections[indexPath.section].rows[indexPath.row]
+
+        let cell = tableView.dequeueReusableCell(withIdentifier: row.reuseIdentifier, for: indexPath)
+        configure(cell, for: row, at: indexPath)
+
+        return cell
+    }
+}
+
+// MARK: - UITableViewDelegate Conformance
+//
+extension ProductCatalogVisibilityViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+        let row = sections[indexPath.section].rows[indexPath.row]
+        if let catalogVisibility = row.catalogVisibility {
+            productSettings.catalogVisibility = catalogVisibility
+        }
+        tableView.reloadData()
+    }
+
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return UITableView.automaticDimension
+    }
+
+    func tableView(_ tableView: UITableView, estimatedHeightForHeaderInSection section: Int) -> CGFloat {
+        return CGFloat.leastNormalMagnitude
+    }
+
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
```

Comment:
from the screenshot, there doesn't seem to be a use of header in the table view. maybe we can just not implement this or return an empty view here?

Reply:
I just removed it 👍